From 3baf6177ea8d01a973d0cbc3fd66c776e1419877 Mon Sep 17 00:00:00 2001
From: Steve Kondik <shade@chemlab.org>
Date: Sun, 24 Nov 2013 21:40:09 -0800
Subject: [PATCH 06/17] recovery: Puke out an /etc/fstab so stuff like
 busybox/toybox is happy

 * And disregard special mount flags on purpose because of certain
   dubious packages which "exec busybox mount".

Squashed:
Author: Alessandro Astone <ales.astone@gmail.com>
Date:   2021-03-26 18:02:34 +0100

    roots: Correct mount flags in /etc/fstab

    The logic here got inverted.

    Change-Id: I63c0d77fa17442ae8630015a52c7bf1be66f69f4

Change-Id: I163702c9bd7fca3d40676fd6d8476e8deb13acc0

recovery: make /etc/fstab only include entries that match the detected fs type

 * toybox's `mount` does not support multiple entries like we do,
   so if we can detect the filesystem of an fstab enrtry print that
   and only that one to /etc/fstab, so that mounting via toybox has
   a better chance of succeding.

 * as a bonus, this patch also gets rid of duplicates in /etc/fstab
   caused by the fact that ReadDefaultFstab() combines entries from
   DT and from recovery.fstab

Change-Id: Iec4ab38044054555d2a33da6f5d53de7716e7bee
---
 recovery_utils/Android.bp |  4 +++
 recovery_utils/roots.cpp  | 52 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 56 insertions(+)

diff --git a/recovery_utils/Android.bp b/recovery_utils/Android.bp
index a48ce002f..34bd588c6 100644
--- a/recovery_utils/Android.bp
+++ b/recovery_utils/Android.bp
@@ -48,6 +48,10 @@ cc_defaults {
         "libhealthhalutils",
         "libhealthshim",
     ],
+
+    whole_static_libs: [
+        "libext2_blkid",
+    ],
 }
 
 // A utility lib that's local to recovery (in contrast, libotautil is exposed to device-specific
diff --git a/recovery_utils/roots.cpp b/recovery_utils/roots.cpp
index 86c7feb57..48a0df639 100644
--- a/recovery_utils/roots.cpp
+++ b/recovery_utils/roots.cpp
@@ -33,6 +33,8 @@
 #include <android-base/properties.h>
 #include <android-base/stringprintf.h>
 #include <android-base/unique_fd.h>
+#include <blkid/blkid.h>
+#include <cryptfs.h>
 #include <ext4_utils/ext4_utils.h>
 #include <ext4_utils/wipe.h>
 #include <fs_mgr.h>
@@ -44,10 +46,39 @@ using android::fs_mgr::Fstab;
 using android::fs_mgr::FstabEntry;
 using android::fs_mgr::ReadDefaultFstab;
 
+static void write_fstab_entry(const FstabEntry& entry, FILE* file) {
+  if (entry.fs_type != "emmc" && !entry.fs_mgr_flags.vold_managed && !entry.blk_device.empty() &&
+      entry.blk_device[0] == '/' && !entry.mount_point.empty() && entry.mount_point[0] == '/') {
+    fprintf(file, "%s ", entry.blk_device.c_str());
+    fprintf(file, "%s ", entry.mount_point.c_str());
+    fprintf(file, "%s ", entry.fs_type.c_str());
+    fprintf(file, "%s 0 0\n", !entry.fs_options.empty() ? entry.fs_options.c_str() : "defaults");
+  }
+}
+
 static Fstab fstab;
 
 constexpr const char* CACHE_ROOT = "/cache";
 
+FstabEntry* fstab_entry_for_mount_point_detect_fs(const std::string& path) {
+  FstabEntry* found = android::fs_mgr::GetEntryForMountPoint(&fstab, path);
+  if (found == nullptr) {
+    return nullptr;
+  }
+
+  if (char* detected_fs_type = blkid_get_tag_value(nullptr, "TYPE", found->blk_device.c_str())) {
+    for (auto& entry : fstab) {
+      if (entry.mount_point == path && entry.fs_type == detected_fs_type) {
+        found = &entry;
+        break;
+      }
+    }
+    free(detected_fs_type);
+  }
+
+  return found;
+}
+
 void load_volume_table() {
   if (!ReadDefaultFstab(&fstab)) {
     LOG(ERROR) << "Failed to read default fstab";
@@ -61,14 +92,35 @@ void load_volume_table() {
       .length = 0,
   });
 
+  Fstab fake_fstab;
   std::cout << "recovery filesystem table" << std::endl << "=========================" << std::endl;
   for (size_t i = 0; i < fstab.size(); ++i) {
     const auto& entry = fstab[i];
     std::cout << "  " << i << " " << entry.mount_point << " "
               << " " << entry.fs_type << " " << entry.blk_device << " " << entry.length
               << std::endl;
+
+    if (std::find_if(fake_fstab.begin(), fake_fstab.end(), [entry](const FstabEntry& e) {
+          return entry.mount_point == e.mount_point;
+        }) == fake_fstab.end()) {
+      FstabEntry* entry_detectfs = fstab_entry_for_mount_point_detect_fs(entry.mount_point);
+      if (entry_detectfs == &entry) {
+        fake_fstab.emplace_back(entry);
+      }
+    }
   }
   std::cout << std::endl;
+
+  // Create a boring /etc/fstab so tools like Busybox work
+  FILE* file = fopen("/etc/fstab", "w");
+  if (file) {
+    for (auto& entry : fake_fstab) {
+      write_fstab_entry(entry, file);
+    }
+    fclose(file);
+  } else {
+    LOG(ERROR) << "Unable to create /etc/fstab";
+  }
 }
 
 Volume* volume_for_mount_point(const std::string& mount_point) {
-- 
2.47.1

